contracts/contract_call.go

```diff
@@ -0,0 +1,116 @@
+// Method represents a contract's method
+type Method struct {
+	abi    *abi.ABI
+	method string
+	maxGas uint64
+}
+
+// NewMethod creates a new contract message
+func NewMethod(abi *abi.ABI, method string, maxGas uint64) Method {
+	return Method{
+		abi:    abi,
+		method: method,
+		maxGas: maxGas,
+	}
+}
+
+func (am Method) Bind(contractAddress common.Address) *BoundMethod {
+	return &BoundMethod{
+		Method:         am,
+		resolveAddress: noopResolver(contractAddress),
+	}
+}
+
+// encodeCall will encodes the msg into []byte format for EVM consumption
+func (am Method) encodeCall(args ...interface{}) ([]byte, error) {
+	return am.abi.Pack(am.method, args...)
+}
+
+// decodeResult will decode the result of msg execution into the result parameter
+func (am Method) decodeResult(result interface{}, output []byte) error {
+	if result == nil {
+		return nil
+	}
+	return am.abi.Unpack(result, am.method, output)
+}
+
+func NewBoundMethod(contractAddress common.Address, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return NewMethod(abi, methodName, maxGas).Bind(contractAddress)
+}
+
+func NewRegisteredContractMethod(registryId common.Hash, abi *abi.ABI, methodName string, maxGas uint64) *BoundMethod {
+	return &BoundMethod{
+		Method: NewMethod(abi, methodName, maxGas),
+		resolveAddress: func(vmRunner vm.EVMRunner) (common.Address, error) {
+			return resolveAddressForCall(vmRunner, registryId, methodName)
+		},
+	}
+}
+
+type BoundMethod struct {
+	Method
+	resolveAddress func(vm.EVMRunner) (common.Address, error)
+}
+
+func (bm *BoundMethod) Query(vmRunner vm.EVMRunner, result interface{}, args ...interface{}) error {
+	return bm.run(vmRunner, result, true, nil, args...)
+}
+
+func (bm *BoundMethod) Execute(vmRunner vm.EVMRunner, result interface{}, value *big.Int, args ...interface{}) error {
+	return bm.run(vmRunner, result, false, value, args...)
+}
+
+func (bm *BoundMethod) run(vmRunner vm.EVMRunner, result interface{}, readOnly bool, value *big.Int, args ...interface{}) error {
+	if vmRunner == nil {
+		return ErrNoEVMRunner
+	}
+
+	defer meterExecutionTime(bm.method)()
+
+	contractAddress, err := bm.resolveAddress(vmRunner)
+	if err != nil {
+		return err
+	}
+
+	logger := log.New("to", contractAddress, "method", bm.method, "args", args, "maxgas", bm.maxGas)
+
+	input, err := bm.encodeCall(args...)
+	if err != nil {
+		logger.Error("Error invoking evm function: can't encode method arguments", "err", err)
+		return err
+	}
+
+	var output []byte
+	var leftoverGas uint64
+	if readOnly {
+		output, leftoverGas, err = vmRunner.Query(contractAddress, input, bm.maxGas)
+	} else {
+		output, leftoverGas, err = vmRunner.Execute(contractAddress, input, bm.maxGas, value)
+	}
+
+	if err != nil {
+		message, _ := unpackError(output)
+		logger.Error("Error invoking evm function: EVM call failure", "input", hexutil.Encode(input), "err", err, "message", message)
+		return err
+	}
+
+	if err := bm.decodeResult(result, output); err != nil {
```

Comment:
Maybe we could return an error here if readOnly is set and result is nil

Reply:
that would change the previous API, so i wouldn't risk it